rbc/heavydb/row_function_manager.py

```diff
@@ -57,121 +60,88 @@ def getDictId(self, func_name: str, arg_idx: int) -> int:
+    get_dict_id_ = external(defn)
 
+    def impl(mgr, func_name, arg_idx):
+        func_name_ = global_str_constant("row_mgr_func_name", func_name)
+        return get_dict_id_(as_voidptr(mgr), func_name_, arg_idx)
+    return impl
 
-@extending.intrinsic
-def heavydb_udf_manager_get_string_dict_proxy_(typingctx, mgr, db_id, dict_id):
-    from .string_dict_proxy import StringDictionaryProxyNumbaType
-    dict_proxy = StringDictionaryProxyNumbaType()
-    sig = dict_proxy(mgr, db_id, dict_id)
 
+@extending.overload_method(RowFunctionManagerNumbaType, 'getDictDbId')
+def heavydb_udf_manager_get_db_id(mgr, func_name, arg_idx):
     target_info = TargetInfo()
     if target_info.software[1][:3] < (6, 2, 0):
         raise UnsupportedError(error_msg % (".".join(map(str, target_info.software[1]))))
 
-    def codegen(context, builder, signature, args):
-        mgr_ptr, db_id, dict_id = args
-
-        mgr_i8ptr = builder.bitcast(mgr_ptr, i8p)
-        fnty = ir.FunctionType(i8p, [i8p, i32, i32])
-        fn = cgutils.get_or_insert_function(
-            builder.module, fnty, "RowFunctionManager_getStringDictionaryProxy")
-        dict_id = builder.trunc(dict_id, i32)
-        db_id = builder.trunc(db_id, i32)
-        proxy_ptr = builder.call(fn, [mgr_i8ptr, db_id, dict_id])
-        proxy_ctor = cgutils.create_struct_proxy(sig.return_type)
-        proxy = proxy_ctor(context, builder)
-        proxy.ptr = proxy_ptr
-        return proxy._getvalue()
-
-    return sig, codegen
-
+    defn = 'int32 RowFunctionManager_getDictDbId(int8*, int8*, int32)'
+    get_dict_db_id_ = external(defn)
 
-@extending.overload_method(HeavyDBRowFunctionManagerNumbaType, 'getDictId')
-def heavydb_udf_manager_get_dict_id(mgr, func_name, arg_idx):
     def impl(mgr, func_name, arg_idx):
-        return heavydb_udf_manager_get_dict_id_(mgr, __DICT_ID, func_name, arg_idx)
+        func_name_ = global_str_constant("row_mgr_func_name", func_name)
+        return get_dict_db_id_(as_voidptr(mgr), func_name_, arg_idx)
     return impl
 
 
-@extending.overload_method(HeavyDBRowFunctionManagerNumbaType, 'getDictDbId')
-def heavydb_udf_manager_get_db_id(mgr, func_name, arg_idx):
-    def impl(mgr, func_name, arg_idx):
-        return heavydb_udf_manager_get_dict_id_(mgr, __DB_ID, func_name, arg_idx)
-    return impl
+@extending.overload_method(RowFunctionManagerNumbaType, 'getStringDictionaryProxy')
+def heavydb_udf_manager_get_string_dict_proxy(mgr, db_id, dict_id):
+    target_info = TargetInfo()
+    if target_info.software[1][:3] < (6, 2, 0):
+        raise UnsupportedError(error_msg % (".".join(map(str, target_info.software[1]))))
 
+    i8p = nb_types.CPointer(nb_types.int8)
+    i32 = nb_types.int32
+    proxy = string_dict_proxy.StringDictionaryProxyNumbaType()
+    sig = typing.signature(proxy, i8p, i32, i32)
+
+    symbol = 'RowFunctionManager_getStringDictionaryProxy'
+    get_string_dict_proxy_ = nb_types.ExternalFunction(symbol, sig)
```

Comment:
Below we have
```python
def heavydb_column_getString(proxy, string_id):
    getBytes = external('int8* StringDictionaryProxy_getStringBytes(int8*, int32)')
    getBytesLength = external('int64 StringDictionaryProxy_getStringLength(int8*, int32)')
    ...
```
for instance. The use of `external` looks cleaner. Can we use it here?

Reply:
In this particular case, no. A string dictionary proxy cannot be an argument of a UDF/UDTF and thus, not initialized. Attempting to do `external('int8* func(StringDictionaryProxy*')` will actually create a new Numba type inheriting from `numba.type.Type`. This was one of the corner cases I had to deal when implementing this function.